Guide/file-storage.markdown

```diff
@@ -575,3 +575,244 @@ config = do
+
+We start by adding the `cryptonite` package to the `default.nix` file, and re-start `./start` to get the package.
+
+```nix
+# default.nix
+haskellEnv = import "${ihp}/NixSupport/default.nix" {
+    ihp = ihp;
+    haskellDeps = p: with p; [
+        cabal-install
+        base
+        # ...
+        cryptonite # <--- ADD THIS
+    ];
+```
+
+Then we need to generate a private and public key pair. We can do it in our Config file, and save the keys for later use.
+
+```haskell
+-- Config/Config.hs
+import Crypto.PubKey.RSA as RSA
+
+newtype RsaPublicAndPrivateKeys = RsaPublicAndPrivateKeys (RSA.PublicKey, RSA.PrivateKey)
+
+config :: ConfigBuilder
+config = do
+    -- See https://ihp.digitallyinduced.com/Guide/config.html
+    -- for what you can do here
+
+    -- ...
+
+    -- Private and public keys to sign and verify image style URLs.
+    -- The 300 means how large your URL can be.
+    -- The 65537 is the public exponent of RSA.
+    -- You can probably leave those values as is.
+    (publicKey, privateKey) <- liftIO $ liftIO $ RSA.generate 300 65537
+    option $ RsaPublicAndPrivateKeys (publicKey, privateKey)
+```
+
+Your Controller will now look like this:
+
+```haskell
+module Web.Controller.ImageStyle where
+
+import Web.Controller.Prelude
+import IHP.ControllerSupport
+import System.Directory (doesFileExist)
+import qualified Data.Text as T
+
+-- Imports for the signed token.
+import Crypto.PubKey.RSA.PKCS15 as RSA
+import Crypto.Hash.Algorithms as Hash.Algorithms
+import Config
+import Data.ByteString.Base64 as Base64
+
+instance Controller ImageStyleController where
+    action RenderImageStyleAction { width, height, originalImagePath, signed } = do
+        let size = show width <> "x" <> show height
+
+        -- Verify the signed token.
+        let Config.RsaPublicAndPrivateKeys (publicKey, _) = getAppConfig @Config.RsaPublicAndPrivateKeys
+
+        -- Verify the token, and deny or allow access based on the result.
+        accessDeniedUnless case cs signed |> Base64.decode of
+            Left msg -> False
+            Right signed -> RSA.verify (Just Hash.Algorithms.SHA256) publicKey (cs $ originalImagePath <> size) signed
```

Comment:
Maybe this can be extracted into it's own function  and then used à la `accessDeniedUnless (signatureMatches imageUrl signature)`

Reply:
Done